File: 18주차/YH/floodFill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        
        h, w = len(image), len(image[0])
        
        visited = set()
		
        def DFS( r, c, src_color, new_color) :
            
            # 색상 변경을 할 수 없는 경우
            # 가로 세로를 초과하는 경우
            # 이미 방문한 경우
            # 기준점의 image[sr][sc]와 값이 다른 경우
            if r < 0 or c < 0 or r >= h or c >= w or (r,c) in visited or image[r][c] != src_color : return
            
            # 예외 값이 아닌 경우는 색상 변경
            image[r][c] = new_color
            
            # 방문한 것 추가
            visited.add( (r,c) )
            
            # 4 방향 모두 검사
            DFS( r-1, c, src_color, new_color )
            DFS( r+1, c, src_color, new_color )
            DFS( r, c-1, src_color, new_color )
            DFS( r, c+1, src_color, new_color )
            
        
        DFS(sr, sc, src_color = image[sr][sc], new_color = newColor)
        
        return image

File: 18주차/SG/Flood_Fill.cpp
#include <bits/stdc++.h>
using namespace std;
int dy[4] = {0,1,0,-1};
int dx[4] = {1,0,-1,0};
class Solution
{
    public:
    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int newColor)
    {
        int startColor = image[sr][sc];
        if(startColor == newColor) return image;
        queue<pair<int,int>> q;
        q.push({sr,sc});
        image[sr][sc] = newColor;
        while(!q.empty()){
            int y = q.front().first;
            int x = q.front().second;
            q.pop();
            for(int i=0;i<4;i++){
                int ny = y+dy[i];
                int nx = x+dx[i];
                if(ny < 0 || ny >= image.size() || nx < 0 || nx >=image[0].size()) continue;
                if(startColor != image[ny][nx]) continue;
                q.push({ny,nx});
                image[ny][nx] = newColor;
            }
        }
        return image;
    }
};